### DISCUSSION TITLE
decoder-rtx

### DISCUSSION THREAD

**Discussion by chrisjmccormick** [chrisjmccormick/stacks#2] [Ideas]
Thread for discussing the [decoder-rtx](https://github.com/chrisjmccormick/stacks/blob/main/stacks/decoder-rtx/README.md) challenge.

**Comment by chrisjmccormick** [comment]
Lots of ideas to try!

---

### Refining Existing Work

**Length Cap Curriculum**
- The curriculum schedule hasn't been tuned
- Adjusting the schedule to step in increments of 64 tokens will turn the current sawtooth pattern into steps.

**Skipping Gradients**
- The VE gates are another target worth trying.
- Smear gate and lambda would make sense, but they're intertwined with the input embedding gradient, so I'm not sure what will happen there.

**Bigram Init**
- Currently the loss increases at the start of training--haven't looked at why yet.

---

### modded-nanogpt

Stuff to try pulling over next:
- Attention gates
- Bigram embeddings
- Memory layouts:
    - Vertically stacked QKVO
        - Then a 'view' of these as a single giant stack of 44 matrices for Muon will reduce those four passes into one large one.
    - MLP W_in and W_out both vertical and stacked in a single bank (so one Muon pass instead of two)
- Run AdamW every other step
- 

---

### Backward Math

- **Clip pre-gradient** - Instead of clipping gradients, which we do after the learning signal has been averaged over the tokens (and presumably drowned out by whatever yields the gradient spike?), can we clip the activations or gradient stream? Maybe that would rescue the step instead of just muting it.
- **Loss scale magnitude** - It's interesting to me that we divide by such a large number at the start of the backward pass, and then further use these small learning rates. Seems risky numerically?
    - Current modded-nanogpt doesn't actually scale the loss at all! (there's a bug)
- **Fuse Optimizer in Backward** - I'm curious if there are any fusion opportunities with the optimizer math and the gradient calculations for, e.g., the LM head. 
    - The main reason to suspect 'maybe not' is that you need the gradients for both momentum buffers--you need grad and grad^2.
- **Re-use activations, ala PPO Clip** - Try taking the activations from some point--maybe the backout layer, maybe the last layer, ...--do one backward pass and update gradients and weights up to that point, then re-use those activations and forward through the rest of the model a second time. Use PPO clip ratio to silence any tokens which get over-confident, and then do a full backward pass.
